File: web/src/presence.ts
import type { ThemeId } from './themes'

export type UserStatus =
  | 'Online' | 'Away' | 'Do Not Disturb' | 'Invisible'
  /** Skype 3-only in the picker — "I'll take calls from strangers." Hidden
   *  from selection starting Skype 4, removed completely by Skype 5. See
   *  docs/skype-eras/skype3.md. */
  | 'Skype Me'
  /** Skype 4-only in the picker — arrived in 4.x, gone again by Skype 5.
   *  See docs/skype-eras/skype4.md. */
  | 'Not Available'

/** The baseline four, present in every era — used as a fallback and by
 *  callers that don't have a theme in scope. */
export const STATUSES: UserStatus[] = ['Online', 'Away', 'Do Not Disturb', 'Invisible']

/** Which self-selectable statuses a given era's status picker should
 *  offer. Both era-specific statuses are added onto, never replacing,
 *  the baseline four — Skype only ever accumulated within any one
 *  release's lifetime; these two are unusual in being removed later, but
 *  neither removal happened within the span of a single theme. */
export function statusesForEra(theme: ThemeId): UserStatus[] {
  if (theme === 'skype3') return [...STATUSES, 'Skype Me']
  if (theme === 'skype4') return [...STATUSES, 'Not Available']
  return STATUSES
}

export const IDLE_MS = 10 * 60 * 1000

// Idle only ever downgrades Online → Away. A status the user picked by
// hand (Away, DND, Invisible, Skype Me, Not Available) sticks until they
// change it.
export function effectiveStatus(manual: UserStatus, idle: boolean): UserStatus {
  return manual === 'Online' && idle ? 'Away' : manual
}

File: nexus_server/presence.go
package main

// The statuses a user can pick. "Offline" is never set directly — it's
// what everyone else sees when you disconnect or go Invisible.
//
// "Skype Me" and "Not Available" are era-specific: real Skype had "Skype
// Me" (an invite for calls from strangers) from its early days, hid it
// from the status picker starting with version 4, and removed it
// completely by version 5; "Not Available" arrived in version 4 and was
// also gone by version 5 (see docs/skype-eras/skype3.md and skype4.md).
// The client only offers these for the eras that actually had them
// (web/src/presence.ts's statusesForEra), but the server accepts both
// unconditionally rather than re-deriving era rules from a username —
// it has no reliable way to know which theme a client is currently
// showing, and rejecting a status the client legitimately offered would
// be a worse failure mode than accepting one from a client that (by a
// bug elsewhere) offered it for the wrong era.
var settableStatuses = map[string]bool{
	"Online":         true,
	"Away":           true,
	"Do Not Disturb": true,
	"Invisible":      true,
	"Skype Me":       true,
	"Not Available":  true,
}

func validStatus(s string) bool {
	return settableStatuses[s]
}

// publicStatus is what friends are told. Invisible users look offline;
// an unset status reads as Online.
func publicStatus(s string) string {
	switch s {
	case "Invisible":
		return "Offline"
	case "":
		return "Online"
	}
	return s
}

// announcePresence loads the user's persisted status, mirrors it onto the
// live connection, and tells their friends. Used at connect/login time in
// place of the old hardcoded "Online" broadcast.
func (s *NexusServer) announcePresence(username string) {
	st := "Online"
	s.DB.QueryRow("SELECT COALESCE(status, 'Online') FROM users WHERE username = ?", username).Scan(&st)
	// Status is mirrored onto every one of the user's connections: it's a
	// property of the person, not of whichever tab happens to be newest.
	s.Mu.Lock()
	for _, c := range s.Clients[username] {
		c.Status = st
	}
	s.Mu.Unlock()
	s.broadcastPresence(username, st)
}
